File: PerfectBinaryTrees/examples/perfect_binary_tree_example.py
from PerfectBinaryTrees.source.perfect_binary_tree import PerfectBinaryTree  # Import the PerfectBinaryTree class

if __name__ == "__main__":
    # Create a Perfect Binary Tree of depth 3
    depth = 3  # Tree will have 2^3 - 1 = 7 nodes
    perfect_tree = PerfectBinaryTree(depth)

    # Print the tree using pre-order traversal
    print("Pre-order Traversal of Perfect Binary Tree:")
    perfect_tree.print_tree(perfect_tree.root)  # Outputs: 1 2 4 5 3 6 7

    # Search for a value in the tree
    search_value1 = 4
    print(f"\nSearching for value {search_value1} in the tree:")
    is_found1 = perfect_tree.search(perfect_tree.root, search_value1)
    print("Found!" if is_found1 else "Not found.")

    # Search for a value not present in the tree
    search_value2 = 10
    print(f"Searching for value {search_value2} in the tree:")
    is_found2 = perfect_tree.search(perfect_tree.root, search_value2)
    print("Found!" if is_found2 else "Not found.")

File: PerfectBinaryTrees/examples/perfect_binary_tree_example.cpp
#include "../include/perfect_binary_tree.h"
#include <iostream>
#include <vector>

int main() {
    // Example data to populate the tree
    std::vector<int> data = {1, 2, 3, 4, 5, 6, 7};

    // Define the depth of the tree
    int depth = 3; // A depth-3 tree will have 7 nodes (if fully filled)

    // Create a PerfectBinaryTree object
    PerfectBinaryTree tree(data, depth);

    // Print the tree using in-order traversal
    std::cout << "In-order Traversal of Perfect Binary Tree: ";
    tree.printTree(tree.root);
    std::cout << std::endl;

    // Search for a value in the tree
    int target = 5;
    if (tree.search(tree.root, target)) {
        std::cout << "Value " << target << " found in the tree!" << std::endl;
    } else {
        std::cout << "Value " << target << " not found in the tree!" << std::endl;
    }

    // Try searching for a value that doesn't exist
    target = 10;
    if (tree.search(tree.root, target)) {
        std::cout << "Value " << target << " found in the tree!" << std::endl;
    } else {
        std::cout << "Value " << target << " not found in the tree!" << std::endl;
    }

    return 0;
}
